Bottom: e7347a264f4017d1c1f335f53a9e44f2d1a69db7
Top:    c8cec6da8a6ce308e9033afb5587f09e5f6d10b6
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-03 10:33:39 +0100

Refresh of give-the-hids-control-over-att

---

diff --git a/src/action.c b/src/action.c
index 622711f..a0577e4 100644
--- a/src/action.c
+++ b/src/action.c
@@ -4038,24 +4038,27 @@ ActionMarkCrosshair (int argc, char **argv, int x, int y)
     {
       if (Marked.status)
 	{
-	  DrawMark ();
+	  notify_mark_change (false);
 	  Marked.status = false;
+	  notify_mark_change (true);
 	}
       else
 	{
+	  notify_mark_change (false);
+	  Marked.status = false;
 	  Marked.status = true;
 	  Marked.X = Crosshair.X;
 	  Marked.Y = Crosshair.Y;
-	  DrawMark ();
+	  notify_mark_change (true);
 	}
     }
   else if (GetFunctionID (function) == F_Center)
     {
-      DrawMark ();
+      notify_mark_change (false);
       Marked.status = true;
       Marked.X = Crosshair.X;
       Marked.Y = Crosshair.Y;
-      DrawMark ();
+      notify_mark_change (true);
     }
   return 0;
 }
diff --git a/src/crosshair.c b/src/crosshair.c
index ca51f4f..c7f7554 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -683,6 +683,15 @@ notify_crosshair_change (bool changes_complete)
     gui->notify_crosshair_change (changes_complete);
 }
 
+
+void
+notify_mark_change (bool changes_complete)
+{
+  if (gui->notify_mark_change)
+    gui->notify_mark_change (changes_complete);
+}
+
+
 /* ---------------------------------------------------------------------------
  * switches crosshair on
  */
diff --git a/src/crosshair.h b/src/crosshair.h
index d413bbe..1135f0e 100644
--- a/src/crosshair.h
+++ b/src/crosshair.h
@@ -47,6 +47,7 @@
 #define	STATE_THIRD		2
 
 void notify_crosshair_change (bool changes_complete);
+void notify_mark_change (bool changes_complete);
 void DrawAttached (void);
 void CrosshairOn (void);
 void CrosshairOff (void);
diff --git a/src/hid.h b/src/hid.h
index 1990f6e..7779f84 100644
--- a/src/hid.h
+++ b/src/hid.h
@@ -299,6 +299,7 @@ typedef enum
     void (*invalidate_lr) (int left_, int right_, int top_, int bottom_);
     void (*invalidate_all) (void);
     void (*notify_crosshair_change) (bool changes_complete);
+    void (*notify_mark_change) (bool changes_complete);
 
     /* During redraw or print/export cycles, this is called once per
        layer (or layer group, for copper layers).  If it returns false
